rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -8,17 +8,258 @@ use crate::{
+        });
+
+        rebuilt.expect("impossible for union to reduce its subtypes to zero")
+    }
+
+    /// Attempt to apply collapse and sorts to intersections, unions, records
+    pub(crate) fn collapse(self) -> Result<Type, CollapseError> {
+        match self {
+            Type::Intersection(intersection) => {
+                let (a, b) = intersection.as_ref();
+                let (a, b) = match (a.clone().collapse(), b.clone().collapse()) {
+                    (Ok(a), Ok(b)) => (a, b),
+                    (Err(a), Err(b)) => return Err(CollapseError::join(vec![a, b])),
+                    (Ok(_), Err(b)) => return Err(b),
+                    (Err(a), Ok(_)) => return Err(a),
+                };
+
+                if a == b {
+                    return Ok(a);
+                }
+
+                match (&a, &b) {
+                    // all intersections should have been collapsed
+                    (Type::Intersection(_), _) => Err(CollapseError(vec![
+                        "intersection could not exist after a collapse".into(),
+                    ])),
+                    (_, Type::Intersection(_)) => Err(CollapseError(vec![
+                        "intersection could not exist after a collapse".into(),
+                    ])),
+                    // intersection of records
+                    (Type::Atom(TypeAtom::Record(a)), Type::Atom(TypeAtom::Record(b))) => {
+                        let mut errors = vec![];
+                        let mut conflicts = BTreeMap::<&Label, BTreeSet<Type>>::new();
+                        let mut fields = BTreeMap::<&Label, Type>::new();
+                        a.iter()
+                            .chain(b.iter())
+                            .for_each(|(label, ty)| match ty.clone().collapse() {
+                                Ok(ty) => {
+                                    if !errors.is_empty() {
+                                        return;
+                                    }
+
+                                    if let Some(ejected) = fields.insert(label, ty.clone()) {
+                                        if ty != ejected {
+                                            conflicts
+                                                .entry(label)
+                                                .and_modify(|set| {
+                                                    set.insert(ty.clone());
+                                                })
+                                                .or_insert(BTreeSet::from([ejected, ty]));
+                                        }
+                                    }
+                                }
+                                Err(err) => errors.push(err),
+                            });
+
+                        if !errors.is_empty() {
+                            return Err(CollapseError::join(errors));
+                        }
+
+                        if !conflicts.is_empty() {
```

Comment:
Shouldn't the errors here be checked alongside `errors`? As a user I prefer receiving all possible errors at once instead of "unlocking them"

Reply:
when there is a collapse error, conflict detection is no longer reliable and is skipped. this is why Err from errors comes first before conflicts